### ISSUE TITLE
Support for debugpy, the successor to python-ptvsd

### ISSUE THREAD

**Issue by leungbk** [emacs-lsp/dap-mode#306]
`python-ptvsd` appears to be deprecated (https://github.com/microsoft/ptvsd/tree/deprecate). It would be nice if support for the new `debugpy` existed in `dap-mode`.

Something like the following communicates with `debugpy`, but I'm pretty sure it's not completely correct.

``` diff
diff --git a/dap-python.el b/dap-python.el
index 09ced1f..c8272e2 100644
--- a/dap-python.el
+++ b/dap-python.el
@@ -190,7 +190,7 @@ as the pyenv version then also return nil. This works around https://github.com/
          (module (plist-get conf :module)))
 
     (plist-put conf :program-to-start
-               (format "%s%s -m ptvsd --wait --host %s --port %s %s %s %s"
+               (format "%s%s -m debugpy --listen %s:%s --wait-for-client %s %s %s"
                        (or dap-python-terminal "")
                        (shell-quote-argument python-executable)
                        host
@@ -215,7 +215,7 @@ as the pyenv version then also return nil. This works around https://github.com/
          (module (plist-get conf :module)))
 
     (plist-put conf :program-to-start
-               (format "%s%s -m ptvsd --wait --host %s --port %s %s %s %s"
+               (format "%s%s -m debugpy --listen %s:%s --wait-for-client %s %s %s"
                        (or dap-python-terminal "")
                        (shell-quote-argument python-executable)
                        host
@@ -238,7 +238,7 @@ as the pyenv version then also return nil. This works around https://github.com/
                                    :cwd nil
                                    :module nil
                                    :program nil
-                                   :request "launch"
+                                   :request "attach"
                                    :name "Python :: Run file (buffer)"))
 
 (dap-register-debug-template "Python :: Run pytest (buffer)"
-- 
2.27.0
```

If `"launch"` is used as the `request`, `debugpy` throws [here](https://github.com/microsoft/debugpy/blob/9ba20c40af2e233b5880dc7e102db18bf8fa3a6e/src/debugpy/adapter/clients.py#L251), expecting an `"attach"` instead.

While `"attach"` is OK for remote debugging, I ran my tests locally. Given that `dap-mode` has a working ``launch`` request for the other supported debuggers, I suspect it should be possible to use `"launch"` here as well.

[Vimspector seems to use `launch`](https://github.com/puremourning/vimspector#python). They feature a test using `"attach"` for a remote setting, but I can't immediately tell if the other `python` test they have uses `"attach"`.

**Comment by yyoncho** [comment]
FTR this is what I do when I am unable to figure out why debug adapter does not work: https://github.com/emacs-lsp/dap-mode/issues/184#issuecomment-572418864

**Comment by leungbk** [comment]
It seems that `debugpy` doesn't support `attach` requests at present: https://github.com/microsoft/debugpy/issues/335

**Comment by nbfalcon** [comment]
It seems to support it now - the Issue was closed. I'll be working on this, since I need debugpy asap to test dap's upcoming :console support.

**Comment by nbfalcon** [comment]
I have implemented preliminary support for debugpy in my "debugpy" branch. However, it does not start, erroring with a connection refused error, so I would appreciate any help with getting that to work. cc @yyoncho

**Comment by yyoncho** [comment]
1. Check if vscode is starting that same process and how it is connecting to it.
2. Check if vscode is sending the same initialization options. You could do that following the instructions at https://github.com/emacs-lsp/dap-mode/issues/184#issuecomment-572418864

**Comment by nbfalcon** [comment]
After some "reverse-engineering" of VSCode, I could conclude that the debugpy adapter is not in fact -m debugpy, as one would expect, but its undocumented sibling, -m debugpy.adapter. I was successfully able to make dap-python work with debugpy (which is much simpler) and will merge into master soon. @yyoncho: currently, dap-python supports both debuggers, configurable with a defcustom and, optionally, per launch configuration. Do you think we should make debugpy the default, potentially breaking lots of configurations if people didn't install debugpy, or keep ptvsd the default, making debugpy opt-in?

**Comment by yyoncho** [comment]
We could register new templates for debugpy and let the user select them if they prefer the new debugger. Breaking changes should be performed only as a last resort. 

In general, dap-mode should fail more gracefully when a particular adapter is not present. ATM this is not implemented. In case it is not present, we should offer to install it when the option is available or point to the readme when it is not.

**Comment by yyoncho** [comment]
We could also create a new file for the new adapter. Later, when we are good with debugpy we might put a warning in dap-python.el that the adapter is deprecated and the users should switch to the new debug adapter.

**Comment by nbfalcon** [comment]
I thought of checking for the respective adapter's availability, too. However, that comes at a cost: overhead. A new python process has to be started and the module loaded, which takes time. As for writing new templates: that isn't necessary. The debugger doesn't change any parameters (only introduces cool new ones), so the only thing that the user needs to do is to opt-in:
```emacs-lisp
(setq dap-python-debugger 'debugpy)
```
and to install the debugger.

**Comment by nbfalcon** [comment]
Regarding making a new file: the entire existing dap-python infrastructure (dap-python--test-at-point) can be used as-is with debugpy. Making a new module would involve code duplication. I kind of still want it to be the "python" debugger, since that is what VSCode uses in launch.json files ("type": "python" -> debugpy).

**Comment by nbfalcon** [comment]
Closing this, since debugpy support is in master.

**Comment by yyoncho** [comment]
> A new python process has to be started and the module loaded, which takes time

This could be performed only once.

> Regarding making a new file: the entire existing dap-python infrastructure (dap-python--test-at-point) can be used as-is with debugpy.

90% of the code in dap-debug.py should be rewritten either way using the lsp-mode helpers.

> I kind of still want it to be the "python" debugger, since that is what VSCode uses in launch.json files ("type": "python" -> debugpy).

I provided a suggestion on how to achieve that without breaking users config.

**Comment by leungbk** [comment]
> Closing this, since debugpy support is in master.

Thanks for working on this. How can I get this working for standalone Python files? When I try running `dap-debug`, I get `Warning (emacs): Initialize request failed: Invalid message: "module" must be str`

When I try providing the `module` in addition to the `program`, I get `Warning (emacs): Initialize request failed: Invalid message: "program", "module", and "code" are mutually exclusive`.

**Comment by leungbk** [comment]
Ah, never mind. It seems that it is not enough to have `(:module nil) in the `conf` plist returned by `dap-python--populate-start-file-args`; the `:module` key must be there at all if we send a non-nil `:program` to `debugpy`. If we do that and make some additional changes to the default template, everything works in `debugpy`.

**Comment by nbfalcon** [comment]
Keys must not be nil. Unlike ptvsd, debugpy is started without arguments; all parameters are provided in keys the launch configuration, which means that all of its keys must either be unspecified or of the correct type. I do think that dap-python needs to be refactored, though.

**Comment by lccambiaghi** [comment]
Can you please post the debug template I can use to run debugpy? Thanks a lot!

**Comment by nbfalcon** [comment]
First, enable debugpy:
```emacs-lisp
(setq dap-python-debugger 'debugpy)
```
(you can also do this in individual templates by adding :debugger 'debugpy)

And thats it. You can use your normal python debug templates, except that all VSCode debugpy options are supported. If you or someone else on your team uses VSCode with launch.json files, you should be able to use them, too, unmodified.

**Comment by nbfalcon** [comment]
By "normal python templates" I mean the ones you were already using, unmodified.

**Comment by lccambiaghi** [comment]
I have set the variable in my config and my previous template still uses ptvsd.

I also tried to set the debugger in the template but still using ptvsd.

Here is my template for reference:
```
(dap-register-debug-template "dap-debug-script"
                               (list :type "python"
                                     :cwd (lsp-workspace-root)
                                     :request "attach"
                                     :debugger 'debugpy
                                     :name "dap-debug-script"))
```

**Comment by leungbk** [comment]
> I have set the variable in my config and my previous template still uses ptvsd.
> 
> I also tried to set the debugger in the template but still using ptvsd.
> 
> Here is my template for reference:

Try this:

``` diff
 dap-python.el | 5 ++---
 1 file changed, 2 insertions(+), 3 deletions(-)

diff --git a/dap-python.el b/dap-python.el
index 3d96541..b96d084 100644
--- a/dap-python.el
+++ b/dap-python.el
@@ -217,6 +217,7 @@ overriden in individual `dap-python' launch configurations."
          (plist-put conf :hostName host)
          (plist-put conf :host host)))
       ('debugpy
+       (plist-put conf :cwd (lsp-workspace-root))
        (plist-put conf :dap-server-path
                   (list python-executable "-m" "debugpy.adapter"))))
     (plist-put conf :program program)
@@ -233,9 +234,7 @@ overriden in individual `dap-python' launch configurations."
 (dap-register-debug-provider "python" 'dap-python--populate-start-file-args)
 (dap-register-debug-template "Python :: Run file (buffer)"
                              (list :type "python"
-                                   :args ""
-                                   :cwd nil
-                                   :module nil
+                                   :args []
                                    :program nil
                                    :request "launch"
                                    :name "Python :: Run file (buffer)"))
-- 
2.28.0

```

**Comment by lccambiaghi** [comment]
I now get: `Wrong type argument: stringp, nil`

This is the debugger output:
```
  funcall-interactively(dap-debug (:type "python" :args [] :request "launch" :name "Python :: Run file (buffer)" :dap-server-path ("/Users/luca/git/wondercast/empties/.venv/bin/python" "-m" "debugpy.adapter") :program "/Users/luca/git/wondercast/empties/src/empties/bookings_feature/merge_bookings_with_activities/__main__.py"))
  #<subr call-interactively>(dap-debug nil nil)
```
It seems that the params fail to be passed to the `dap-debug` function?

**Comment by leungbk** [comment]
> I now get: `Wrong type argument: stringp, nil`
> 
> This is the debugger output:
> 
> ```
>   funcall-interactively(dap-debug (:type "python" :args [] :request "launch" :name "Python :: Run file (buffer)" :dap-server-path ("/Users/luca/git/wondercast/empties/.venv/bin/python" "-m" "debugpy.adapter") :program "/Users/luca/git/wondercast/empties/src/empties/bookings_feature/merge_bookings_with_activities/__main__.py"))
>   #<subr call-interactively>(dap-debug nil nil)
> ```
> 
> It seems that the params fail to be passed to the `dap-debug` function?

The `cwd` isn't getting passed because you didn't run `M-x eval-defun` on the function I suggested modifying.

**Comment by nbfalcon** [comment]
@Iccambiaghi: don't use nil anywhere in your template, as debugpy doesn't handle it (leave it out instead). Aside from that, what makes you assume that debugpy isn't used? Have you looked at your task manager to ascertain that ptvsd is being used? There isn't any noticable (ui-related) difference between the two. Have you updated?

**Comment by lccambiaghi** [comment]
I don't have `nil` in my template and I copy pasted your changes to `dap-python.el`.. Weird! 

@nbfalcon I can see the command in the bottom window. The command is:
```
/.../.venv/bin/python -m ptvsd --wait --host localhost --port 53399  /.../__main__.py 
```

Maybe it is something with my `lsp` setup or in general `doom-emacs` config? Thank you anyway for your help!

**Comment by nbfalcon** [comment]
@lccambiaghi are you sure that you have updated to latest dap-mode? If so, try restarting doom. I'd really like this issue resolved.

**Comment by nbfalcon** [comment]
Of course! Doom pins packages! You might not have the latest dap-mode installed.

**Comment by nbfalcon** [comment]
@lccambiaghi the solution is probably to add (package! dap-mode) to your $DOOMDIR/packages.el and to run doom/sync. This will give you the latest version from MELPA, with this feature.

**Comment by lccambiaghi** [comment]
Does the latest version include the fix? I can see the last release is on 6th of August?

**Comment by nbfalcon** [comment]
There is no "fix" - debugpy support was implemented at some point, and your version of dap-mode might be too old. By updating you get the feature.

**Comment by nbfalcon** [comment]
I just added a bunch of key nil checks in latest master, so that the built-in templates actually work with debugpy now.

**Comment by lccambiaghi** [comment]
You were right, it was somehow my Doom Emacs stuck on an old release. I can confirm I can run `debugpy` now. Looking forward to the console support :)

**Comment by nbfalcon** [comment]
@lccambiaghi there is a release on MELPA. All features are in master and were implemented more than 3 hours ago, which means that they are already built. The issue with doom is that, for the sake of reproducibility and stability, it does not pull directly from MELPA, but instead downloads pinned releases. This means that doom will not magically break, but you won't get the latest software either.
